### ISSUE TITLE
TrustKit crashes intermittently

### ISSUE THREAD

**Issue by frederoni** [mapbox/mapbox-events-ios#9]
TrustKit crashes intermittently on this [line](https://github.com/mapbox/mapbox-events-ios/blob/master/vendor/TrustKit/Reporting/reporting_utils.m#L23) 
It seems to happen more frequently when pushing new map view's into the navigation stack.
For example when going from overview into a turn-by-by navigation in the [mapbox/mapbox-navigation-ios](https://github.com/mapbox/mapbox-navigation-ios) example but not exclusively.

<details>
<pre>
Thread 14 Queue : com.datatheorem.trustkit.reporterqueue (serial)
#0	0x0000000102bf24e0 in dispatch_sync_f ()
#1	0x0000000182320974 in SecTrustGetVerifyTime ()
#2	0x0000000182320a98 in SecTrustEvaluateIfNecessary ()
#3	0x0000000182321bd0 in SecTrustGetCertificateCount ()
#4	0x0000000101bdb870 in convertTrustToPemArray at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/Reporting/reporting_utils.m:20
#5	0x0000000101be1eb4 in -[TSKPinningValidatorResult certificateChain] at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TSKPinningValidatorResult.m:24
#6	0x0000000101bdd898 in -[TrustKit sendValidationReport:notedHostname:pinningPolicy:] at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TrustKit.m:215
#7	0x0000000101bdd3d8 in __46-[TrustKit initWithConfiguration:isSingleton:]_block_invoke_2 at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TrustKit.m:175
#8	0x0000000101be1840 in __49-[TSKPinningValidator evaluateTrust:forHostname:]_block_invoke at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TSKPinningValidator.m:204
#9	0x0000000102bed69c in _dispatch_call_block_and_release ()
#10	0x0000000102bed65c in _dispatch_client_callout ()
#11	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#12	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#13	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#14	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#15	0x00000001812e3fcc in _pthread_wqthread ()
#16	0x00000001812e3c1c in start_wqthread ()
Enqueued from  (Thread 26) Queue :  (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x0000000101be16e0 in -[TSKPinningValidator evaluateTrust:forHostname:] at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TSKPinningValidator.m:203
#2	0x0000000101be1b3c in -[TSKPinningValidator handleChallenge:completionHandler:] at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/vendor/TrustKit/TSKPinningValidator.m:224
#3	0x0000000101bd643c in -[MMENSURLSessionWrapper URLSession:didReceiveChallenge:completionHandler:] at /Users/fredrik/Workspace/mapbox/voyage-ios/Pods/MapboxMobileEvents/MapboxMobileEvents/MMENSURLSessionWrapper.m:40
#4	0x000000018207d7c8 in __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ ()
#5	0x0000000181fbf034 in -[NSBlockOperation main] ()
#6	0x0000000181faec14 in -[__NSOperationInternal _start:] ()
#7	0x0000000102bed65c in _dispatch_client_callout ()
#8	0x0000000102bf9d98 in _dispatch_block_invoke_direct ()
#9	0x0000000102bed65c in _dispatch_client_callout ()
#10	0x0000000102bf9d98 in _dispatch_block_invoke_direct ()
#11	0x0000000102bf9c58 in dispatch_block_perform ()
#12	0x000000018207f4a0 in __NSOQSchedule_f ()
#13	0x0000000102bed65c in _dispatch_client_callout ()
#14	0x0000000102bfaa24 in _dispatch_continuation_pop ()
#15	0x0000000102bf92c0 in _dispatch_async_redirect_invoke ()
#16	0x0000000102bfecc0 in _dispatch_root_queue_drain ()
#17	0x0000000102bfe9ec in _dispatch_worker_thread3 ()
#18	0x00000001812e411c in _pthread_wqthread ()
#19	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.NSURLSession-work (Thread 22) Queue : com.apple.NSURLSession-work (serial)
#0	0x0000000102bfcbb4 in _dispatch_root_queue_push ()
#1	0x000000018207c80c in __NSOQSchedule ()
#2	0x000000018207e4d0 in __addOperations ()
#3	0x0000000181dea470 in -[NSURLSession delegate_didReceiveChallenge:completionHandler:] ()
#4	0x0000000181c5a69c in -[__NSCFLocalSessionTask connection:challenged:authCallback:] ()
#5	0x0000000181cd7d24 in __56-[__NSCFURLLocalSessionConnection _didReceiveChallenge:]_block_invoke.397 ()
#6	0x0000000181cd7a50 in -[__NSCFURLLocalSessionConnection _didReceiveChallenge:] ()
#7	0x0000000181d996b0 in ___ZN19URLConnectionLoader41protocolDidReceiveAuthenticationChallengeEP19_CFURLAuthChallenge_block_invoke_3 ()
#8	0x0000000181cd9610 in ___ZN25SessionConnectionLoadable21withLoaderClientAsyncEU13block_pointerFvP21LoaderClientInterfaceE_block_invoke ()
#9	0x0000000102bed69c in _dispatch_call_block_and_release ()
#10	0x0000000102bed65c in _dispatch_client_callout ()
#11	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#12	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#13	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#14	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#15	0x00000001812e3fcc in _pthread_wqthread ()
#16	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.NSURLSession-work (Thread 22) Queue : com.apple.NSURLSession-work (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x0000000181cd3f44 in -[__NSCFURLSessionConnection withWorkQueueAsync:] ()
#2	0x0000000181cd8d28 in SessionConnectionLoadable::withLoaderClientAsync(void (LoaderClientInterface*) block_pointer) ()
#3	0x0000000181d99580 in URLConnectionLoader::protocolDidReceiveAuthenticationChallenge(_CFURLAuthChallenge*) ()
#4	0x0000000181d9f45c in ___ZN27URLConnectionLoader_Classic41protocolDidReceiveAuthenticationChallengeEP19_CFURLAuthChallenge_block_invoke ()
#5	0x0000000181d8d554 in ___ZNK25URLConnectionInstanceData18withWorkQueueAsyncEU13block_pointerFvvE_block_invoke ()
#6	0x0000000102bed69c in _dispatch_call_block_and_release ()
#7	0x0000000102bed65c in _dispatch_client_callout ()
#8	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#9	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#10	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#11	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#12	0x00000001812e3fcc in _pthread_wqthread ()
#13	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.CFNetwork.LoaderQ (Thread 26) Queue : com.apple.CFNetwork.LoaderQ (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x0000000181d8d524 in URLConnectionInstanceData::withWorkQueueAsync(void () block_pointer) const ()
#2	0x0000000181d9f420 in URLConnectionLoader_Classic::protocolDidReceiveAuthenticationChallenge(_CFURLAuthChallenge*) ()
#3	0x0000000181d56088 in HTTPProtocol::needServerTrust(__SecTrust*, void (SSLPeerTrustAcceptancePolicy, bool) block_pointer) ()
#4	0x0000000181d3c2a8 in ___ZN4Tube24_onqueue_needServerTrustEv_block_invoke_2 ()
#5	0x0000000181e64b3c in ___ZNK18QCoreSchedulingSet12performAsyncEU13block_pointerFvvE_block_invoke ()
#6	0x0000000102bed69c in _dispatch_call_block_and_release ()
#7	0x0000000102bed65c in _dispatch_client_callout ()
#8	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#9	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#10	0x0000000102bfc0fc in _dispatch_queue_serial_drain ()
#11	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#12	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#13	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#14	0x00000001812e3fcc in _pthread_wqthread ()
#15	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.CFNetwork.Connection (Thread 26) Queue : com.apple.CFNetwork.Connection (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x0000000181e64a70 in QCoreSchedulingSet::performAsync(void () block_pointer) const ()
#2	0x0000000181d3c208 in Tube::_onqueue_needServerTrust() ()
#3	0x0000000181dc7678 in TCPIOConnection::_tlsProcessValidateServerCertificates(nw_tls_context*) ()
#4	0x0000000181dc5b0c in TCPIOConnection::_handleTLSEvent(nw_tls_context*, tls_handshake_message_t, void () block_pointer, bool) ()
#5	0x00000001825999cc in __tcp_connection_set_tls_message_handler_on_nw_connection_block_invoke ()
#6	0x000000018247618c in __boringssl_context_handshake_message_handler_block_invoke ()
#7	0x0000000102bed69c in _dispatch_call_block_and_release ()
#8	0x0000000102bed65c in _dispatch_client_callout ()
#9	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#10	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#11	0x0000000102bfc0fc in _dispatch_queue_serial_drain ()
#12	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#13	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#14	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#15	0x00000001812e3fcc in _pthread_wqthread ()
#16	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.network.connections (Thread 26) Queue : com.apple.network.connections (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x000000018246e5dc in boringssl_context_handshake_message_handler ()
#2	0x0000000182484d84 in ssl3_connect ()
#3	0x00000001824db2f4 in SSL_do_handshake ()
#4	0x000000018246d9d4 in boringssl_session_handshake_continue ()
#5	0x000000018246d4c0 in boringssl_context_handshake_negotiate ()
#6	0x00000001824783e0 in nw_protocol_boringssl_handshake_negotiate ()
#7	0x0000000102bed69c in _dispatch_call_block_and_release ()
#8	0x0000000102bed65c in _dispatch_client_callout ()
#9	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#10	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#11	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#12	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#13	0x00000001812e3fcc in _pthread_wqthread ()
#14	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.CFNetwork.Connection (Thread 26) Queue : com.apple.CFNetwork.Connection (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x00000001824761f0 in __boringssl_context_handshake_message_handler_block_invoke_2 ()
#2	0x0000000181dc5a6c in TCPIOConnection::_handleTLSEvent(nw_tls_context*, tls_handshake_message_t, void () block_pointer, bool) ()
#3	0x00000001825999cc in __tcp_connection_set_tls_message_handler_on_nw_connection_block_invoke ()
#4	0x000000018247618c in __boringssl_context_handshake_message_handler_block_invoke ()
#5	0x0000000102bed69c in _dispatch_call_block_and_release ()
#6	0x0000000102bed65c in _dispatch_client_callout ()
#7	0x0000000102bfc2dc in _dispatch_queue_serial_drain ()
#8	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#9	0x0000000102bfc0fc in _dispatch_queue_serial_drain ()
#10	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#11	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#12	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#13	0x00000001812e3fcc in _pthread_wqthread ()
#14	0x00000001812e3c1c in start_wqthread ()
Enqueued from com.apple.network.connections (Thread 26) Queue : com.apple.network.connections (serial)
#0	0x0000000102bfce1c in _dispatch_queue_push ()
#1	0x000000018246e5dc in boringssl_context_handshake_message_handler ()
#2	0x0000000182484d84 in ssl3_connect ()
#3	0x00000001824db2f4 in SSL_do_handshake ()
#4	0x000000018246d9d4 in boringssl_session_handshake_continue ()
#5	0x000000018246d4c0 in boringssl_context_handshake_negotiate ()
#6	0x00000001824783e0 in nw_protocol_boringssl_handshake_negotiate ()
#7	0x0000000182478884 in nw_boringssl_read ()
#8	0x000000018247993c in nw_protocol_boringssl_input_available ()
#9	0x0000000182572910 in nw_socket_update_input_source ()
#10	0x0000000102bed65c in _dispatch_client_callout ()
#11	0x0000000102bfaa24 in _dispatch_continuation_pop ()
#12	0x0000000102bef458 in _dispatch_source_invoke ()
#13	0x0000000102bfc0fc in _dispatch_queue_serial_drain ()
#14	0x0000000102bf0bf0 in _dispatch_queue_invoke ()
#15	0x0000000102bfd2b8 in _dispatch_root_queue_drain_deferred_wlh ()
#16	0x0000000102c04268 in _dispatch_workloop_worker_thread ()
#17	0x00000001812e3fcc in _pthread_wqthread ()
#18	0x00000001812e3c1c in start_wqthread ()
</pre>
</details>

@boundsj 👀

**Comment by boundsj** [comment]
TrustKit appears to have a bug that can cause a crash when:

a) certificate validation fails for some reason
b) the reporting feature of TrustKit is used to report the failure

We did not intend to use the reporting feature of TrustKit at this time and will never use the default reporting URL. I disabled in https://github.com/mapbox/mapbox-events-ios/commit/d1626cd36ca4fb83ed399e045613357bd55e007b and released that change in [v0.2.9](https://github.com/mapbox/mapbox-events-ios/releases/tag/v0.2.9) (also published on CocoaPods)

Updating to v0.2.9 of MapboxMobileEvents should work around this issue. We will need to file a new issue upstream to fix the actual issue. Closing here.
